### PR TITLE
Libssh1 and libssh2 test

### PR DESCRIPTION
**Pull request by tonyyuan1**
Libssh1 and libssh2 test

- Related ticket: https://progress.opensuse.org/issues/73420
- Needles: no
- Verification run: 
sles15sp3: http://10.67.183.130/tests/131
sles15sp2: http://10.67.183.130/tests/125
sles15sp1: http://10.67.183.130/tests/132
sles12sp5: http://10.67.183.130/tests/121
sles12sp4: http://10.67.183.130/tests/47
opensuse-Tumbleweed: http://10.67.183.130/tests/126

Detailed test cases: http://bugzilla.suse.com/tr_show_case.cgi?case_id=1768668

### CODE REVIEW

File: tests/console/libssh.pm [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    zypper_call("in docker libvirt-daemon-qemu qemu-kvm sshpass qemu-block-ssh");
+    systemctl("start docker.service libvirtd.service");
+    create_image;
+    assert_script_run("iptables -I INPUT 5 -p tcp --dport 22 -j ACCEPT");
+    assert_script_run("mkdir /tmp/test; echo -n libssh_testcase001 > /tmp/test/libssh_testfile");
+    assert_script_run("qemu-img create -f raw /tmp/test/libssh_block.raw 20m");    #file to be uploaded to container for qemu block test
+    assert_script_run('docker run -itd --name libssh_container --hostname container01 --add-host susetest:$(ip -4 addr show docker0 | grep -Po "inet \K[\d.]+") --privileged libssh_image');
+
+    #start container as test client
+    type_string("docker exec -it libssh_container bash\n", wait_still_screen => 3);
```
Comment (grisu48) [inline tests/console/libssh.pm:87]: Not sure what `wait_still_screen` does in `root-console` mode. Is this necessary?

**Review by grisu48 — APPROVED** [review]
LGTM! Perhaps you consider also porting this test to openSUSE TW/Leap?

File: tests/console/libssh.pm (hunk 1 above)
Comment (tonyyuan1) [reply tests/console/libssh.pm:87]: > Not sure what `wait_still_screen` does in `root-console` mode. Is this necessary?

All other api like "script_run" need send back the shell command result "$?" via ttyS0 but "docker exec" will take the test inside container. The container will take over the ttyS0 .  ttyS0 will not be give back to host until the test inside container is done. It's not possible to know how much time it take to give ttyS0 back to host. It likely cause timeout if script_run is used.

**Comment by tonyyuan1** [comment]
> LGTM! Perhaps you consider also porting this test to openSUSE TW/Leap?

Yes. but maybe need to build opensuse container. opensuse container image tag format seems different. I will try.

**Comment by jouyingbin** [comment]
@tonyyuan1 Thanks Tony for the libssh test, the PR is LGTM, I just have some questions.
1. Will the case be able to run in our original test before, or it is only working under container base testing.
2. Will it also supports our original FIPS test too?

**Comment by tonyyuan1** [comment]
> LGTM! Perhaps you consider also porting this test to openSUSE TW/Leap?

Updated the code. It covers openSUSE TW/Leap and sle15sp3 now

**Comment by tonyyuan1** [comment]
> @tonyyuan1 Thanks Tony for the libssh test, the PR is LGTM, I just have some questions.
> 
>     1. Will the case be able to run in our original test before, or it is only working under container base testing.
It's a normal test but using container technology a bit. 
> 
>     2. Will it also supports our original FIPS test too?
Yes, it does but need a bit more setup before turning system into Fips mode. We should check with Marcus if it's needed to test libssh FIPS

**Comment by tonyyuan1** [comment]
@foursixnine please review.

**Review by foursixnine — APPROVED** [review]
Aside from the header (wrong year) Looks pretty good!, also kudos for thinking about opensuse aswell!

File: tests/console/libssh.pm [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+# The container will be used as client of libssh
+# The variable: dev_image_tag can be used to specify a base image different from host OS. If the base images is
+# stored in internel repository i.e registry.suse.de we assume the host OS is in develping stage i.e currently sle15sp3:
+# registry.suse.de/suse/sle-15-sp3/update/cr/totest/images/suse/sle15:15.3
+#
+sub create_image {
+    my $tag = get_var("DEV_IMAGE_TAG");
+    my $kubic;
+    if ($tag =~ /registry\.suse\.de/) {
+        assert_script_run("curl http://ca.suse.de/certificates/ca/SUSE_Trust_Root.crt -o /etc/pki/trust/anchors/SUSE_Trust_Root.crt");
```
Comment (pdostal) [inline tests/console/libssh.pm:49]: You can use `utils::ensure_ca_certificates_suse_installed`

File: tests/console/libssh.pm [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    assert_script_run(qq(docker build -t libssh_image --build-arg kubic="$kubic" --build-arg tag="$tag" --build-arg pkgs="$pkgs" .), timeout => 600);
+}
+
+sub run {
+    select_console 'root-console';
+    unless (is_opensuse) {
+        is_sle('<15') ? add_suseconnect_product("sle-module-containers", 12) : add_suseconnect_product("sle-module-containers");
+    }
+    # Host is used as server of libssh test
+    zypper_call("--gpg-auto-import-keys in docker libvirt-daemon-qemu qemu-kvm qemu-block-ssh");
```
Comment (pdostal) [inline tests/console/libssh.pm:102]: You can also simplify this test module by using `containers::common::install_docker_when_needed` a lot.

File: tests/console/libssh.pm [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    validate_script_output("curl -s sftp://susetest/tmp/test/libssh_testfile -u root:nots3cr3t", sub { m/libssh_testcase001/ });
+    assert_script_run("curl -s sftp://susetest/tmp/test/libssh_block.raw -u root:nots3cr3t -o /tmp/libssh_block.raw");
+    validate_script_output("curl -s scp://susetest/tmp/test/libssh_testfile -u root:nots3cr3t",             sub { m/libssh_testcase001/ });
+    validate_script_output('curl -s sftp://root@susetest/tmp/test/libssh_testfile --key /root/.ssh/id_rsa', sub { m/libssh_testcase001/ });
+    validate_script_output('curl -s scp://root@susetest/tmp/test/libssh_testfile --key /root/.ssh/id_rsa',  sub { m/libssh_testcase001/ });
+    validate_script_output('virsh -c "qemu+libssh://root@susetest/system?sshauth=privkey&keyfile=/root/.ssh/id_rsa&known_hosts=/root/.ssh/known_hosts" hostname', sub { m/susetest/ }) if is_sle('>=15'); #libssh is not supported by libvirt for sle12
+    validate_script_output('virsh -c "qemu+libssh2://root@susetest/system?sshauth=privkey&keyfile=/root/.ssh/id_rsa&known_hosts=/root/.ssh/known_hosts" hostname', sub { m/susetest/ });
```
Comment (pdostal) [inline tests/console/libssh.pm:130]:
```suggestion
    if is_sle('>=15') { #libssh is not supported by libvirt for sle12
        validate_script_output('virsh -c "qemu+libssh://root@susetest/system?sshauth=privkey&keyfile=/root/.ssh/id_rsa&known_hosts=/root/.ssh/known_hosts" hostname', sub { m/susetest/ });
    } else {
        validate_script_output('virsh -c "qemu+libssh2://root@susetest/system?sshauth=privkey&keyfile=/root/.ssh/id_rsa&known_hosts=/root/.ssh/known_hosts" hostname', sub { m/susetest/ });
    }
```

File: tests/console/libssh.pm [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+info block virtio1
+quit
+EOF
+', sub { m/libssh_block\.raw/ });
+
+    assert_script_run("docker stop libssh_container");
+}
+
+sub test_flags {
+    return {always_rollback => 1};
```
Comment (pdostal) [inline tests/console/libssh.pm:147]: Why is this in here? This means that this test won't work for example on bare metal or on public cloud.

**Review by pdostal — CHANGES_REQUESTED** [review]
This is very good job Tony! I'm sorry I'm so late with my review. I hope those suggestions won't delay you too much but I'd like to try to do some code deduplication over here.

**Comment by tonyyuan1** [comment]
> This is very good job Tony! I'm sorry I'm so late with my review. I hope those suggestions won't delay you too much but I'd like to try to do some code deduplication over here.

Nothing is delayed. The quality always comes first. Thank you very much for review!

File: tests/console/libssh.pm (hunk 2 above)
Comment (tonyyuan1) [reply tests/console/libssh.pm:49]: Will use ensure_ca_certificates_suse_installe.

File: tests/console/libssh.pm (hunk 3 above)
Comment (tonyyuan1) [reply tests/console/libssh.pm:102]: Will use install_docker_when_needed but it seems that it has a bug.  The container extension uri for suseconnect is always 12 (sle-module-containers/12/x86_64). There are not 12.1, 12.2 ..... 
Line 88:  https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/containers/common.pm
add_suseconnect_product("sle-module-containers") will fail to add module. 
add_suseconnect_product("sle-module-containers", 12) should work.

File: tests/console/libssh.pm (hunk 4 above)
Comment (tonyyuan1) [reply tests/console/libssh.pm:130]: The original logic will change:
original logic : "virsh -c "qemu+libssh2" will run for all sle versions.
After change as you suggested: "virsh -c "qemu+libssh2" will not run if is_sle('>=15')

File: tests/console/libssh.pm (hunk 5 above)
Comment (tonyyuan1) [reply tests/console/libssh.pm:147]: This test installed many packages i.e libvirt, qemu ... and created container. It's difficult to clean up by coding,  I used always_rollback to clean up them all so as not to affect or break the tests after it.

Comment (tonyyuan1) [reply tests/console/libssh.pm:147]: I documented testcases for libssh in testopia so reviewers can better understand the test to save time. 
https://bugzilla.suse.com/tr_show_case.cgi?case_id=1768668

File: tests/console/libssh.pm (hunk 4 above)
Comment (pdostal) [reply tests/console/libssh.pm:130]: OK, now I see the difference. Sorry

File: tests/console/libssh.pm (hunk 5 above)
Comment (pdostal) [reply tests/console/libssh.pm:147]: OK. Then just please keep in mind that this won't work on bare metal as well as on public cloud.

**Review by pdostal — APPROVED** [review]

**Review by foursixnine — APPROVED** [review]

**Review by jouyingbin — COMMENTED** [review]
LGTM. Sorry forget to update it. Thanks Tony.

**Pull request merged**